service/lib/agama/software/manager.rb

```diff
@@ -190,6 +193,52 @@ def patterns(filtered)
         patterns
       end
 
+      def add_pattern(id)
+        # TODO: error handling
+        res = Yast::Pkg.ResolvableInstall(id, :pattern)
+        logger.info "Adding pattern #{res.inspect}"
+        @user_patterns << id
+
+        res = Yast::Pkg.PkgSolve(unused = true)
+        logger.info "Solver run #{res.inspect}"
+        selected_patterns_changed
+      end
+
+      def remove_pattern(id)
+        # TODO: error handling
+        Yast::Pkg.ResolvableRemove(id, :pattern)
+        logger.info "Removing pattern #{res.inspect}"
+        @user_patterns << id
+
+        res = Yast::Pkg.PkgSolve(unused = true)
+        logger.info "Solver run #{res.inspect}"
+        selected_patterns_changed
+      end
+
+      def user_patterns=(ids)
+        @user_patterns.each { |p| Yast::Pkg.ResolvableRemove(p, :pattern) }
```

Comment:
Rather use the [Pkg.ResolvableNeutral()](https://github.com/yast/yast-pkg-bindings/blob/a790b6b072d6ead55dcda63dc9102ff0fa336d64/src/Resolvable_Install.cc#L404) call here.

`Pkg.ResolvableRemove()`  means "uninstall the resolvable", while `Pkg.ResolvableNeutral()` means "remove all scheduled transactions, do not touch it".